app/src/main/java/net/gsantner/markor/format/markdown/MarkdownTextConverter.java

```diff
@@ -323,4 +377,62 @@ public boolean isFileOutOfThisFormat(String filepath) {
         filepath = filepath.replace(JavaPasswordbasedCryption.DEFAULT_ENCRYPTION_EXTENSION, "");
         return (MarkdownTextConverter.PATTERN_HAS_FILE_EXTENSION_FOR_THIS_FORMAT.matcher(filepath).matches() && !filepath.toLowerCase().endsWith(".txt")) || filepath.toLowerCase().endsWith(".md.txt");
     }
+
+    private Map<String, List<String>> extractYamlAttributes(final String markup) {
+        Parser yamlFrontmatterParser = Parser.builder().extensions(Collections.singleton(YamlFrontMatterExtension.create())).build();
+        AbstractYamlFrontMatterVisitor visitor = new AbstractYamlFrontMatterVisitor();
+        Node document = yamlFrontmatterParser.parse(markup);
+        visitor.visit(document);
+        return visitor.getData();
+    }
+
+    private String replaceTokens(final String markup, final String scopes, boolean wrapInDiv) {
+        String markupReplaced = markup;
+        String attrVal_S = "";
+        String attrVal_E = "";
+
+        for (Map.Entry<String, List<String>> entry : yamlAttributeMap.entrySet()) {
+            String attrName = entry.getKey();
+            List<String> attrValue = entry.getValue();
+            List<String> attrValueOut = new ArrayList<>();
+
+            if (attrName.equals("tags") && attrValue.size() == 1) {
+                // It's not a real tag list, but rather a string of comma-separated strings
+                attrValue = Arrays.asList(attrValue.get(0).split(",\\s*"));
+            }
+
+            if (wrapInDiv) {
+                attrVal_S = HTML_TOKEN_ITEM_CONTAINER_S;
+                attrVal_E = HTML_TOKEN_ITEM_CONTAINER_E + "\n";
+            } else {
+                attrVal_S = "";
+                attrVal_E = "";
+            }
+            for (String aValue : attrValue) {
+                // Strip surrounding single or double quotes
+                aValue = aValue.replaceFirst("^(['\"])(.*)\\1", "$2");
+                aValue = TextUtils.htmlEncode(aValue);
+                aValue = aValue.replaceAll("`(.*?)`", "<code>$1</code>");
+                aValue = aValue.replaceAll("_(.*?)_", "<em>$1</em>");
+                aValue = aValue.replaceAll("\\*(.*?)\\*", "<b>$1</b>");
+                aValue = aValue.replaceAll("(?<!-)---(?!-)", "&mdash;");
+                aValue = aValue.replaceAll("(?<!-)--(?!-)", "&ndash;");
+                aValue = aValue.replaceAll("\\$(.*?)\\$", "<span class='katex'>$1</span>");
+                attrValueOut.add(HTML_TOKEN_ITEM_S + aValue + HTML_TOKEN_ITEM_E);
+            }
+
+            // Replace "{{ <scope>>.<key> }}" tokens in note body, if they are not escaped with a preceeding backslash
+            for (String scope : scopes.split(",\\s*")) {
+                String tokenPattern = "(?<!\\\\)\\{\\{ " + scope + "\\." + attrName + " \\}\\}";
```

Comment:
don't make it more complex (also performance wise) than it is. simple search + replace is enough.  
`{{ + scope + . + attrname + }}`

![image](https://user-images.githubusercontent.com/6735650/154805817-125d7216-166a-4626-b2a2-94f107199638.png)

Reply:
The idea was to prevent tokens from being replaced: `{{ page.title }}` would expand, while `\{{ page.title }}` would be left untouched.

As an alternative I can move the backslash into the braces, e.g. `{{ \page.title }}`. Then I could get rid of the `tokenPattern` regex and use simple string replacement instead.

Reply:
no, don't add special escape code here.

if post.title is "hello"

and somewhere it is .... \{{ post.title }} it's fine to get \5

Reply:
Sometimes maybe not. Consider the following example. It would be impossible to type this:

``````
---
title: My Jekyll Notes
---

Here's how to use tags in Jekyll:

```
<ul>
  {% for post in site.posts %}
    <li>
      <a href="{{ post.url }}">{{ post.title }}</a>
    </li>
  {% endfor %}
</ul>
```
``````

It would render as:

``````
Here's how to use tags in Jekyll:

```
<ul>
  {% for post in site.posts %}
    <li>
      <a href="{{ post.url }}">My Jekyll Notes</a>
    </li>
  {% endfor %}
</ul>
```
``````

Reply:
this is correct - but it's not what we are goaling for. markor doesn't have any CMS blog software or Liquid builtin. What you post there is mostly about writing i.e. pagination internals for a blog.

what however more the goal is here - that you can write articles with a yaml ----title: "something"--- and you also get rendered that title in the file where it's references.

again - goal isn't a full blown static site generator. keep it simple :smile:

Reply:
No, I mean the literal code example. Assume I'm learning to use Jekyll and make notes of it using Markor. I would write `I can refer to the current document's title using {{ post.title }}.` It would not render correctly.

Reply:
Ah yeah well there is a pretty simple way here - don't specify title in yaml then. Or use invisible space character somewhere there. I prefer to not lose performance by regex look arounds to this very special case - where you want to just describe the use case.

Reply:
Using invisible space is a nice idea, I like it :smile: Will drop the escape stuff.